### PR TITLE
Add SetDefaults for pipelineSpec in reconciler 

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes

This commit adds SetDefaults for pipelineSpec in reconciler
after it is fetched from cluster. This is to avoid failing the resources that
are not mutated and add default values.
# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by Yongxuanzhang** [comment]
https://github.com/tektoncd/pipeline/issues/5175

**Comment by Yongxuanzhang** [comment]
@jerop

File: pkg/apis/pipeline/v1/task_validation.go [hunk 1]
```diff
@@ -274,6 +274,10 @@ func ValidateParameterTypes(ctx context.Context, params []ParamSpec) (errs *apis
 
 // ValidateType checks that the type of a ParamSpec is allowed and its default value matches that type
 func (p ParamSpec) ValidateType() *apis.FieldError {
+	// For resources not call SetDefaults(), empty type should be valid.
```
Comment (afrittoli) [inline pkg/apis/pipeline/v1/task_validation.go:277]: NIT - suggested comment

> A resource may have been created before type was available in `paramSpec` and thus have no type set

Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1/task_validation.go:277]: sure!

Comment (dibyom) [reply pkg/apis/pipeline/v1/task_validation.go:277]: couldn't we call SetDefaults() before calling validation instead of just skipping validation?

File: pkg/apis/pipeline/v1/task_validation.go [hunk 2]
```diff
@@ -274,6 +274,10 @@ func ValidateParameterTypes(ctx context.Context, params []ParamSpec) (errs *apis
...
+	if p.Type == "" {
```
Comment (ywluogg) [inline pkg/apis/pipeline/v1/task_validation.go:278]: Can you add a unit test in pipeline reconciler that the param type is null, to make sure this wouldn't introduce any new bugs? Thanks

File: pkg/apis/pipeline/v1beta1/task_validation_test.go [hunk 3]
```diff
@@ -436,6 +436,48 @@ func TestTaskSpecValidate(t *testing.T) {
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ts := &v1beta1.TaskSpec{
+				Params: tt.fields.Params,
+				Steps:  tt.fields.Steps,
+			}
+			if err := ts.Validate(context.Background()); err != nil {
+				t.Errorf("TaskSpec.Validate() = %v", err)
+			}
+		})
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/task_validation_test.go:477]: In the test, can we compare the returned v.s. expected error like this [example](https://github.com/tektoncd/pipeline/blob/a4f9216e8957e52ce9bdf4af82cfd82148659539/pkg/apis/pipeline/v1beta1/task_validation_test.go#L1304-L1311) rather than just check if err is nil?

File: pkg/apis/pipeline/v1/task_validation.go (hunk 1 above)
Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1/task_validation.go:277]: > couldn't we call SetDefaults() before calling validation instead of just skipping validation?

yeah I would prefer this way!

Comment (dibyom) [reply pkg/apis/pipeline/v1/task_validation.go:277]: (or at least in this function, we should assume no type == defaultType)

**Comment by abayer** [comment]
It's worth mentioning that when I deleted a problem `Pipeline` and recreated it, with any of `kubectl replace`, `kubectl apply`, or `kubectl create`, params without types specified still had no types specified. This is with Pipeline v0.37.2. So if the current behavior, before this PR, is supposed to be defaulting newly created `Pipeline`s and `Task`s to have default types on their params, well, that's not working either.

**Comment by chuangw6** [comment]
> It's worth mentioning that when I deleted a problem `Pipeline` and recreated it, with any of `kubectl replace`, `kubectl apply`, or `kubectl create`, params without types specified still had no types specified. This is with Pipeline v0.37.2. So if the current behavior, before this PR, is supposed to be defaulting newly created `Pipeline`s and `Task`s to have default types on their params, well, that's not working either.

Hi @abayer,
I tried the following yaml (the type of param `arg` and `mystring` is not specified) with v0.37.2 and it works fine. Can you also try that? If this example works, let's find what are the missing/different part in the problem `pipeline`.

```yaml
kind: Task
apiVersion: tekton.dev/v1beta1
metadata:
  name: test-task
spec:
  params:
    - name: arg
  steps:
    - name: echo
      image: bash:latest
      args: ["echo", "$(params.arg)"]
---
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: test-pipeline
spec:
  params:
    - name: mystring
  tasks:
    - name: echo-test
      params:
        - name: arg
          value: $(params.mystring)
      taskRef:
        name: test-task
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: test-pipeline-run
spec:
  params:
    - name: mystring
      value: abc.com
  pipelineRef:
    name: test-pipeline

```

**Comment by abayer** [comment]
Ok, turns out the problem was incredibly stupid and specific - in the `default` namespace on the dogfooding cluster, we had `webhooks.knative.dev/exclude: "true"` in the labels, so it was skipping defaulting entirely. 🤦

**Comment by ywluogg** [comment]
> Ok, turns out the problem was incredibly stupid and specific - in the `default` namespace on the dogfooding cluster, we had `webhooks.knative.dev/exclude: "true"` in the labels, so it was skipping defaulting entirely. 🤦

Haha lol! But this actually helped us catching some bugs in webhook and strengthen our checks in reconciler! Effort not wasted!

**Comment by ywluogg** [comment]
> Ok, turns out the problem was incredibly stupid and specific - in the `default` namespace on the dogfooding cluster, we had `webhooks.knative.dev/exclude: "true"` in the labels, so it was skipping defaulting entirely. 🤦

🤦🤦🤦

**Comment by chuangw6** [comment]
> Ok, turns out the problem was incredibly stupid and specific - in the `default` namespace on the dogfooding cluster, we had `webhooks.knative.dev/exclude: "true"` in the labels, so it was skipping defaulting entirely. 🤦

Thanks for finding out this. I am curious if this field `webhooks.knative.dev/exclude` was set from false to true just recently OR it has been `true` all the time?

**Comment by Yongxuanzhang** [comment]
> It's worth mentioning that when I deleted a problem `Pipeline` and recreated it, with any of `kubectl replace`, `kubectl apply`, or `kubectl create`, params without types specified still had no types specified. This is with Pipeline v0.37.2. So if the current behavior, before this PR, is supposed to be defaulting newly created `Pipeline`s and `Task`s to have default types on their params, well, that's not working either.



> Ok, turns out the problem was incredibly stupid and specific - in the `default` namespace on the dogfooding cluster, we had `webhooks.knative.dev/exclude: "true"` in the labels, so it was skipping defaulting entirely. 🤦

Is it something very easy to fix? We're not so familiar with the dogfooding cluster configuration. Is it public and can be viewed somewhere?

**Comment by abayer** [comment]
Someone set it by hand to be able to remove resources that couldn't be deleted due to removed fields - not sure exactly when it was done, but more than a week or so ago. It's fixed now, so we're all good on dogfooding regarding defaulting.

**Comment by chuangw6** [comment]
> Someone set it by hand to be able to remove resources that couldn't be deleted due to removed fields - not sure exactly when it was done, but more than a week or so ago. It's fixed now, so we're all good on dogfooding regarding defaulting.

Got it. Good lesson to learn. Didn't know there is a field `webhooks.knative.dev/exclude` that can disable the wekbhook in the cluster. Thanks @abayer

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 4]
```diff
@@ -348,6 +348,10 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun, get
 	}
 
 	pipelineMeta, pipelineSpec, err := rprp.GetPipelineData(ctx, pr, getPipelineFunc)
+	if pipelineSpec != nil {
```
Comment (ywluogg) [inline pkg/reconciler/pipelinerun/pipelinerun.go:351]: There should be a test case testing this and should be corresponds to this change?

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/pipelinerun.go:351]: yes I haven't finished all tests yet! Once done I will unhold this pr

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 5]
```diff
@@ -565,6 +565,7 @@ spec:
   serviceAccountName: test-sa
   taskRef:
     name: unit-test-task
+    kind: Task
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:568]: What's the reason to add `kind: Task` to all of these taskRefs?

**Review by lbernick — COMMENTED** [review]
I'm a bit confused about what this PR is doing given the original PR message and the conversation that's happened. It looks like this sets defaults in the reconciler to ensure that defaults are applied to resources that were created before these code changes were applied to the server. Is this still an issue with params without types, disregarding the annotations that were mistakenly kept around on our dogfooding cluster?

**Comment by Yongxuanzhang** [comment]
> I'm a bit confused about what this PR is doing given the original PR message and the conversation that's happened. It looks like this sets defaults in the reconciler to ensure that defaults are applied to resources that were created before these code changes were applied to the server. 

Yes,  that's the current goal

> Is this still an issue with params without types, disregarding the annotations that were mistakenly kept around on our dogfooding cluster?

There are some other discussions in this issue https://github.com/tektoncd/pipeline/issues/5175. And for taskrun when we get the taskspec, it will call `taskSpec.SetDefaults(ctx)`, 
https://github.com/tektoncd/pipeline/blob/10a5ae69b35bd7cc4542234f395c30ca214ca5a6/pkg/reconciler/taskrun/resources/taskspec.go#L48-L54
but for `pipelinerun` we don't have it. It may make sense to add for pipelinerun as well
https://github.com/tektoncd/pipeline/blob/10a5ae69b35bd7cc4542234f395c30ca214ca5a6/pkg/reconciler/pipelinerun/pipelinespec/pipelinespec.go#L40-L47

File: pkg/reconciler/pipelinerun/pipelinerun_test.go (hunk 5 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:568]: setdefaults will add kind for task as well, if not adding them the current tests will fail

**Review by lbernick — APPROVED** [review]
